### ISSUE TITLE
Tensorflow 2.0 tf.name_scope has no effect on weights created by keras functional api

### ISSUE THREAD

**Issue by zzh8829** [tensorflow/tensorflow#27298]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.14
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0-alpha0
- Python version: 3.7.1

**Describe the current behavior**

tf.name_scope does not effect the name of weights created by keras.layers

output in tf 2.0.0-alpha0 
```
dense/kernel:0
dense/bias:0
```

**Describe the expected behavior**
tf.name_scope is applied to wights created by keras.layers
output in tf 1.13.1 is the expected behavior
```
block/dense/kernel:0
block/dense/bias:0
```
**Code to reproduce the issue**
```
import tensorflow as tf
print(tf.__version__)

inputs = tf.keras.Input(shape=[2])
with tf.name_scope('block'):
    outputs = tf.keras.layers.Dense(10)(inputs)
model = tf.keras.Model(inputs, outputs)
for w in model.weights:
    print(w.name)
```

**Other info / logs**

I was able to reproduce the output of tf 1.13 in tf 2.0.0 by using the following code
```
from tensorflow.python.keras.backend import get_graph
with get_graph().as_default(), tf.name_scope('block'):
    outputs = tf.keras.layers.Dense(10)(inputs)
```
keras.layers is using the `keras graph` which overrides the name_scope generated by   tf.name_scope in the `default graph`

I think the problem is actually caused by tf.name_scope incorrectly set `_has_symbolic_input_in_eager` to False when we are building graph using the keras functional api. since the input here is keras.Input, the name_scope should be applied to keras graph instead of default graph

**Comment by gshashank84** [comment]
In TF 2.0 the Eager Execution is set as **Default**, therefore in eager execetion the TF completely ignores the **name_scope**. I think this ops is particularly designed for Graphs.

Therefore you can do all the above things when Graphs are set as Default
```python
with tf.Graph().as_default():  
  inputs = tf.keras.Input(shape=[2])
  with tf.name_scope('block'):
      outputs = tf.keras.layers.Dense(10)(inputs)
model = tf.keras.Model(inputs, outputs)
for w in model.weights:
    print(w.name)
```

**Comment by zzh8829** [comment]
I would disagree here, in a real eager execution scenario, keras layers does work with tf.name_scope. 
```
inputs = np.random.random((1, 2))
with tf.name_scope('block'):
    dense = tf.keras.layers.Dense(10, input_shape=(2, ))
    outputs = dense(inputs)
for w in dense.weights:
    print(w.name)
```
output:
```
block/dense/kernel:0
block/dense/bias:0
```

I don't think keras should have inconsistent behavior when the input type change from numerical to symbolic.  tf.name_scope v1 used to take a `variable` parameter for this exact case, but it's removed in tf2.

In fact, an old [keras tutorial](https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html) specifically mentioned tf.name_scope is compatible with the functional api which is no longer true in tf2.

I don't know if this is indeed the expected outcome, but this definitely needs more clarification since name_scope is pretty essential for building complex models.

**Comment by gshashank84** [comment]
Maybe you're right. Lets see what does the Member of Tensorflow have to say about this.

**Comment by Daniel451** [comment]
Please do not hesitate to add something to my Stackoverflow issue [TensorFlow 2.0: how to group graph using tf.keras? tf.name_scope/tf.variable_scope not used anymore?](https://stackoverflow.com/questions/55318952/tensorflow-2-0-how-to-group-graph-using-tf-keras-tf-name-scope-tf-variable-sco/55376814?noredirect=1#comment97651032_55376814)

However, I also think this is actually a bug. `tf.Variable` works as intended:
```
In [3]: with tf.name_scope("foo"): 
   ...:     with tf.name_scope("bar"): 
   ...:         v = tf.Variable([0])

In [4]: v                                                                                                                                                                                                           
Out[4]: <tf.Variable 'foo/bar/Variable:0' shape=(1,) dtype=int32, numpy=array([0], dtype=int32)>
```
The variable is placed within the scope `foo/bar`.

Eager execution, as alread demonstrated, also works using `tf.keras.layers`...
```
In [5]: inputs = np.random.rand(10,1) 
    ...: with tf.name_scope("foo"): 
    ...:     with tf.name_scope("bar"): 
    ...:         dense_layer = tf.keras.layers.Dense(1) 
    ...:         dense_output = dense_layer(inputs)

In [6]: dense_layer.weights                                                                                                                                                                                        
Out[6]:
[<tf.Variable 'foo/bar/dense_1/kernel:0' shape=(1, 1) dtype=float64, numpy=array([[1.40278891]])>,
 <tf.Variable 'foo/bar/dense_1/bias:0' shape=(1,) dtype=float64, numpy=array([0.])>]
```

...while simply building a model with `tf.keras.layers` completeley ignores `tf.name_scope`:
```
In [7]: with tf.name_scope("foo"): 
    ...:     with tf.name_scope("bar"): 
    ...:         inputs = tf.keras.Input(shape=(1,)) 
    ...:         dense_layer = tf.keras.layers.Dense(1) 
    ...:         dense_output = dense_layer(inputs) 
    ...:         model = tf.keras.Model(inputs=inputs, outputs=dense_output) 
    ...:         model.compile(optimizer=tf.optimizers.SGD(), loss=tf.losses.MeanSquaredError())

In [8]: for l in model.layers: 
    ...:     for w in l.weights: 
    ...:         print(w.name) 
    ...:                                                                                                                                                                                                            
Out[8]: 
dense_9/kernel:0
dense_9/bias:0
```

**Comment by zzh8829** [comment]
^ name_scope is definitely not working very nicely in tf2, throwing in some of my findings here.

In eager mode with symbolic input, tf.name_scope pushes scope into default context here https://github.com/tensorflow/tensorflow/blob/6a603d8cf6df64cb9bd7bbac33245a02821cf198/tensorflow/python/framework/ops.py#L6400

In keras, when layers gets call with symbolic input, it tries to build a graph and replace the current context with the keras graph here https://github.com/tensorflow/tensorflow/blob/6a603d8cf6df64cb9bd7bbac33245a02821cf198/tensorflow/python/keras/engine/base_layer.py#L580

However, if the input is numpy, `__call__` follows this code path 
https://github.com/tensorflow/tensorflow/blob/6a603d8cf6df64cb9bd7bbac33245a02821cf198/tensorflow/python/keras/engine/base_layer.py#L644

which correctly inherits the parent tf.name_scope


The weirdness happens because tf.name_scope is not in the same graph as keras model. I think the best solution might be providing tf.name_scope a extra parameter (like it did before) to specify the input type.

**Comment by cottrell** [comment]
I'm seeing this too ... does anyone know in what direction this is moving?

**Comment by jvishnuvardhan** [comment]
I could reproduce the issue in `tf-nightly-2.0-preview`. Please take a look at the [gist](https://colab.sandbox.google.com/gist/jvishnuvardhan/d517caf07e65b06cb4209c55e1de0025/tf27298.ipynb). Thanks!

**Comment by Daniel451** [comment]
@zzh8829 @rchao are there any updates? Any plans to fix the `name_scope` integration?

**Comment by mharradon** [comment]
I found the name scope worked correctly if I additionally wrapped the `with tf.name_scope(...` with `with tensorflow.python.keras.backend.get_graph().as_default():`. This was sort of mirroring this line: 

https://github.com/tensorflow/tensorflow/blob/6a603d8cf6df64cb9bd7bbac33245a02821cf198/tensorflow/python/keras/engine/base_layer.py#L580

My guess was that the name scopes were specific to eager mode vs graph-based mode, and the normal Keras model construction forced the graph based mode - so you want to specify the name scope in graph mode. This is a very vague guess, and I don't know if this causes other issues.

Using the normal `tensorflow.keras.backend` also didn't work - I had to use `tensorflow.python.keras.backend`.

**Comment by edfall** [comment]
I find similar problem, and here is one solution for name_scope control with Sequential rather than tf.name_scope for tf 2.0 keras：

```python
import tensorflow as tf
print(tf.__version__)

inputs = tf.keras.Input(shape=[2,10])
dense = tf.keras.layers.Dense(10)
seq = tf.keras.Sequential(name='block')
seq.add(dense)
outputs = seq(inputs)
model = tf.keras.Model(inputs, outputs)
for w in model.weights:
    print(w.name)

2.0.0
block/dense_4/kernel:0
block/dense_4/bias:0

```

**Comment by colinrgodsey** [comment]
It works as expected inside the 'build' phase. You're in a functional graph and name_space stack there.

**Comment by Daniel451** [comment]
@jvishnuvardhan @zzh8829 any updates? @rchao does not seem to be active in this issue.

**Comment by ntlex** [comment]
@zzh8829 facing the same issue here. Any ETA on this? Or any case that we can increase the priority of this issue?

**Comment by jvishnuvardhan** [comment]
@zzh8829 I think this was resolved in recent `tf-nightly`. I could not reproduce with `tf-nightly`. [Here](https://colab.research.google.com/gist/jvishnuvardhan/1ece1f2942cfb0aadbd5a72e693e867d/untitled56.ipynb) is the colab gist for your reference.

Please verify once and if this was resolved for you then close the issue. Thanks!

**Comment by jvishnuvardhan** [comment]
@zzh8829 Can you verify once and close the issue if this was resolved for you. Thanks!

**Comment by jvishnuvardhan** [comment]
@zzh8829 I am closing this issue as this was resolved in recent `tf-nightly`. Please feel free to reopen if this issue was not resolved for you. Thanks!

**Comment by Daniel451** [comment]
@jvishnuvardhan this issue not resolved in `2.2.0`, including latest nightly.

Please see this gist https://gist.github.com/Daniel451/03d173825235369be46b27e06e5e109a

**Comment by jvishnuvardhan** [comment]
@Daniel451 Can you please open a new issue with your standalone code. I think the second example in your code uses graph context and providing scope name works well as shown below. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/79327bcdc98a09d3f72afcec18d076c7/untitled109.ipynb).
```
with tf.name_scope("foo2"): 
     with tf.name_scope("block1") as scope: 
         inputs = tf.keras.Input(shape=(1,)) 
         dense_layer = tf.keras.layers.Dense(1,name=scope)(inputs) 
         model = tf.keras.Model(inputs=inputs, outputs=dense_layer) 
         model.compile(optimizer=tf.optimizers.SGD(), loss=tf.losses.MeanSquaredError())

for l in model.layers:
  print(l.name)
  for w in l.weights: 
    print(w.name)
```
Output is 

```
input_2
foo2/block1/
foo2/block1/kernel:0
foo2/block1/bias:0
```
Please open a new issue for further question on this issue. Thanks!

**Comment by phsyron** [comment]
@jvishnuvardhan I can't help but notice that the `tf.name_scope` is a no-op in that example.  The following gives the same output:

```
scope = 'foo2/block1'
inputs = tf.keras.Input(shape=(1,)) 
dense_layer = tf.keras.layers.Dense(1,name=scope)(inputs) 
model = tf.keras.Model(inputs=inputs, outputs=dense_layer) 
model.compile(optimizer=tf.optimizers.SGD(), loss=tf.losses.MeanSquaredError())

for l in model.layers:
	print(l.name)
	for w in l.weights: 
	print(w.name)
```

What we want is for `tf.name_scope` to work, not a tutorial on how to work around it by manually calculating values for `name=`

**Comment by tgsmith61591** [comment]
How is this still not resolved in tf 2.3.0?

**Comment by nicolasshu** [comment]
@tgsmith61591 I have been able to use it since tf 2.1
Have you put it under a `with` block? Like

```
class Douche(tf.keras.Model):
    def __init__(self,out_labels):
        self.shitty = tf.keras.layers.Dense(4)
        self.wok = tf.keras.layers.Dense(out_labels)
    def call(x):
        with tf.name_scope("Turd"):
            x = self.shitty(x)
        with tf.name_scope("Sandwich"):
            x = self.wok(x)
        return x
```

**Comment by Daniel451** [comment]
@nicolasshu this is **not** what this issue is about. The problem persists during graph mode / when constructing the model.

Your example is just another variation of @jvishnuvardhan demonstration that given an actual call / eager mode `name_scope` is inferred correctly. However, as @phsyron already stated, the issue persits for graph mode / model construction -> your example only works, because you subclass `tf.keras.Model` and infer `name_scope` in `call`, i.e. *during* runtime with actual inputs `x` given. We were talking about correctly inferring `name_scope` during model construction and only once as well as a correct inferring during graph mode.

**Comment by Daniel451** [comment]
@jvishnuvardhan I actually lost track of this issue, but it still persists. If opening a new issue is necessary I'll do this and supply a standalone Gist soon. Thank you for your help & time.

**Comment by jvishnuvardhan** [comment]
@Daniel451 If you don't mind, please open a new issue as there are many comments above to follow the actual problem. In the new issue, you can refer this issue for further details. Thanks!

**Comment by kaiyu-pony** [comment]
I was wondering how `tf.keras.layers.Layer` is always able to append the prefix to all the operations inside while `tf.name_scope` doesn't work for `tf.keras.Input`. Looks like the key lies in [here](https://github.com/tensorflow/tensorflow/blob/da6568a/tensorflow/python/keras/engine/base_layer.py#L875-L878).

**Comment by jvishnuvardhan** [comment]
Reopened it as this is still an issue with `tf-nightly` 2.9.0-dev20220222. Thanks!

**Comment by mohantym** [comment]
It is still replicating in [2.9 ](https://colab.sandbox.google.com/gist/mohantym/45b195bfa177a64af960f530308f8ad6/untitled56.ipynb#scrollTo=659SZpY7a5eu)and [2.10.0dev ](https://colab.sandbox.google.com/gist/mohantym/e4c752c98c44fda71e1514a91b4aa441/untitled56.ipynb#scrollTo=S-9S7X1kKXCH).

**Comment by sushreebarsa** [comment]
@zzh8829 I tried to replicate this issue using TF v2.11,and faced the below output. Please find the [gist](https://colab.research.google.com/gist/sushreebarsa/a0acbe789e5f6a4eb0a95423ed85fe2a/untitled56.ipynb#scrollTo=659SZpY7a5eu) and refer to this official [doc](https://www.tensorflow.org/guide/keras/functional).
This issue seems to be Keras issue.  Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues) as Keras development is fully moving to [github.com/keras-team/keras](http://github.com/keras-team/keras). All issues and PRs related to keras will be addressed in that repo.
To know more see this TF forum discussion ; 
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!

**Comment by Harrypotterrrr** [comment]
Sorry for the late reply, as the representive for Google... Though I am not tensorflow/keras developer, but I found the most elegant way to achieve this would be:
```
tf.enable_v2_behavior()
tf.keras.__internal__.apply_name_scope_on_model_declaration(True)
```
to enable the name scope for model declaration using functional API.
